abdiff/core/collate_ab_transforms.py

```diff
@@ -167,89 +173,90 @@ def get_transformed_batches_iter(
     abdiff.core.utils.write_to_dataset() to perform batch writes to
     a Parquet dataset.
     """
+    count = 0
+    total_transformed_files = len(ab_transformed_file_lists[0]) + len(
+        ab_transformed_file_lists[1]
+    )
+
     for transformed_files in ab_transformed_file_lists:
         for transformed_file in transformed_files:
+            count += 1
+            logger.debug(
+                f"Yielding records from file {count} / {total_transformed_files}"
+                f": {transformed_file}"
+            )
             record_iter = get_transformed_records_iter(
                 transformed_file=str(Path(run_directory) / transformed_file)
             )
             for record_batch in itertools.batched(record_iter, READ_BATCH_SIZE):
                 yield pa.RecordBatch.from_pylist(list(record_batch))
 
 
-def get_joined_batches_iter(dataset_directory: str) -> Generator[pa.RecordBatch]:
+def get_joined_batches_iter(
+    dataset_directory: str,
+) -> Generator[pa.RecordBatch, None, None]:
     """Yield pyarrow.RecordBatch objects of joined TIMDEX A/B records.
 
-    This function uses DuckDB to query the Parquet dataset of transformed
-    TIMDEX record dictionaries. It's worth noting that this Parquet dataset
-    is stored in a tempfile.TemporaryDirectory that gets deleted after the
-    function exits.
-
-    The following steps are performed in sequential order:
-        1. A list of DISTINCT transformed filenames (without the file extension)
-           is retrieved.
+    The previous step created a parquet dataset where each A an B version of a record,
+    from a single input file, was a distinct row in the dataset.  This function joins the
+    A and B versions of the same intellectual record as a single row, by limiting to the
+    transformed file.
 
-        Steps 2-4 are performed for each transformed file.
+    For performance reasons, DuckDB is provided with ONLY the parquet files that are
+    associated with that transformed file, to prevent it from scanning all files in the
+    dataset (this was memory safe, but somewhat slow given the potentially large number
+    of transformed files).
 
-        2. Using common table expressions (CTEs) execute a query that performs
-           the following:
-           - Create a CTE called 'transformed_file' that contains all the records
-             where the partition transformed_file_name=the name of the transformed file.
-           - Create a CTE called 'a' that contains all the records from CTE
-             'transformed_file' where the column version='a'.
-           - Create a CTE called 'b' that contains all the records from CTE
-             'transformed_file' where the column version='b'.
-           - Join the 'a' and 'b' CTEs using the column timdex_record_id.
+    As records are joined, an "abdiff_record_id" UUID is minted to unambiguously reference
+    that particular row, which is helpful later during deduping.
```

Comment:
Excellent question; it's not obvious at a glance, or even a medium dive.  

We use DuckDB to mint the UUID [here](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-383-pipeline-tweaks-large-runs/abdiff/core/collate_ab_transforms.py#L265) when we are joining A/B versions of records from an input file.  It's key to point out that this `timdex_record_id` (e.g. `alma:abc123`) could be in _multiple_ input files, potentially from the same day ("full" and "daily"), sometimes "to-index", sometimes "to-delete", etc.  This is why we require the de-duping stage later in the collate process.

As we move into the deduping process, we are faced with the challenge of identifying _one_ instance of an intellectual record (i.e. one A/B combination from a single input file, as records cannot duplicate within a single file.... that we know of).  We know that `timdex_record_id` is not unique across multiple files, and some edge cases have shown that even `timdex_record_id` + `run_date` + `action` may not be unique.  99% of the time that combination would be, but there are some weird edge cases.

When we dedupe, we [perform some ordering and skipping](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-383-pipeline-tweaks-large-runs/abdiff/core/collate_ab_transforms.py#L310-L335).  This addition of `abdiff_record_id` was added as part of this PR, because it was here we ran into out-of-memory (OOM) errors.  The partitioning and grouping by `timdex_record_id` required bringing _all_ rows into memory, such that we could identify the final version.  By adding the `abdiff_record_id` earlier in the process, we can [select _only_ this column during the expensive ordering operation](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-383-pipeline-tweaks-large-runs/abdiff/core/collate_ab_transforms.py#L312), dramatically reducing the memory footprint of this expensive operation.

Then, with this unambiguous identifier to that particular record version, we can [join against the full dataset](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-383-pipeline-tweaks-large-runs/abdiff/core/collate_ab_transforms.py#L346) confident we're getting only unique records.  This last join is "lazy", and is only evaluated as we consume the iterator response.

**In summary**: the addition of `abdiff_record_id` allows for less columns to select during an expensive operation, while still providing a key identifier to join later as we lazily yield those records.

This topic has come a few times over the last few months, why minting a somewhat temporary, internal identifier for ETL data can be helpful.  I think this demonstrates it pretty well.  As we chop and munge and subset data, it avoids the need to reference an intellectual "row" of data by multiple keys, that _may_ result in dupes, by providing a single, unambiguous way to join data.  In this case, it was pretty instrumental for addressing ballooning memory.

Reply:
@jonavellecuerdo - please see the comment below yours (maybe came through around the same time?), thinking perhaps this addresses your question?